### DISCUSSION TITLE
How  to perform data augmentation in model pipeline texcat spacy?

### DISCUSSION THREAD

**Discussion by blizaga** [KennethEnevoldsen/augmenty#203] [General]
I'm trying to follow the tutorial for data augmentation, but it only explains for text and then what about the textcat label on each text. How to do the augmentation?

![image](https://github.com/KennethEnevoldsen/augmenty/assets/54091450/daec8605-c67f-45dc-a503-a6ed71c0b161)

I created a dataframe for text columns and labels as input to the spacy training process using the CLI. Here I use Indonesian language

**Comment by KennethEnevoldsen** [comment]
Hei, sounds like you can just apply the augmentation on the each document independently. 

I am not at a computer at the moment, but I can send a snippet when I am.

**Reply by blizaga** [reply]
How to augment existing labels? Like the famous SMOTE sampling method.

**Reply by KennethEnevoldsen** [reply]
Ahh right. So you want to augment a specific document class/label (e.g. negative). Here is some dummy code to illustrate how you would go about doing so:

Define a function: 

```python
import spacy
import augmenty

nlp = spacy.load("en_core_web_md")

def augment_documents(texts, your_augmenter):
  docs = nlp.pipe(texts)
  augmented_docs = augmenty.docs(docs, augmenter=entity_augmenter, nlp=nlp)
  return [doc.text for doc in augmented_docs] # convert to strings
```

Then utilize that function for the class that you want to create more examples of:

```python
negative_examples = df[df$label == "negative"]["text"].tolist()

# a sample augmenter from augmenty
entity_augmenter = augmenty.load("ents_replace_v1", 
                                 ent_dict = {"ORG": [["spaCy"], ["spaCy", "Universe"]]}, level=1)

new_negative_examples = augment_documents(negative_examples, entity_augmenter)
```

You can now add them to the dataframe with the label "negative" and you will have twice the number of negative samples as before. Of course you can repeat the process with multiple augmenters.

**Reply by blizaga** [reply]
Thanks for the insight, I slightly adjusted your code so that each augmentation iteration can directly produce a "text" column and a "label" column with a negative label.

**Reply by blizaga** [reply]
However, I would like to ask again, how to implement multi augmentation using the function above?

**Reply by KennethEnevoldsen** [reply]
By multiaugmentation I assume you mean creating multiple augmentations from the same text? I.e. for the text: 

> Hope you are well
You might get:

> Hope you are ell
> Hope you well are
> Hope you well 
> hope you well 

One option is to use multiple augmenters. Another is to use the [repeat](https://kennethenevoldsen.github.io/augmenty/augmenty.augment_utilities.html#augmenty.augment_utilities.repeat) augmenter using [combined](https://kennethenevoldsen.github.io/augmenty/augmenty.augment_utilities.html#augmenty.augment_utilities.repeat) augmenters.

**Reply by blizaga** [reply]
Can you provide code examples especially for combination augmentation?

**Reply by KennethEnevoldsen** [reply]
I have added an example in the documentation for using combine, which is as follows:

```python
char_swap_augmenter = augmenty.load("char_swap_v1", level=.02)
synonym_augmenter = augmenty.load("wordnet_synonym_v1", level=1, lang="en")
combined_aug = augmenty.combine([char_swap_augmenter, synonym_augmenter])
# combine doc using two augmenters
augmented_docs = augmenty.docs(docs, augmenter=combined_aug, nlp=nlp)
```

**Reply by blizaga** [reply]
thank you for the insight, it's very useful
